README.md
Mistakes of Significance
================
Nathan (Nat) Goodman
June 12, 2019

<!-- README.md is generated from README.Rmd. Please edit that file -->
A collection of R scripts and documents exploring mistakes made by significance testing. The content at present are

-   blog post [When You Select Significant Findings, You’re Selecting Inflated Estimates](https://replicationnetwork.com/2019/02/16/goodman-when-youre-selecting-significant-findings-youre-selecting-inflated-estimates/) whose main point is that significance testing is a biased procedure that overestimates effect size; also available on my website [here](https://natgoodman.github.io/misig/ovrfx.stable.html)
-   blog post [Your P-Values are Too Small! And So Are Your Confidence Intervals!](https://replicationnetwork.com/2019/05/01/your-p-values-are-too-small-and-so-are-your-confidence-intervals/) which explores the impact of heterogeneous effect sizes on p-values and confidence intervals; also available on my website [here](https://natgoodman.github.io/misig/ovrht.stable.html)
-   supplementary material for the blog posts TBD

**THE SOFTWARE IS A WORK IN PROGRESS. THE CODE IS ROUGH and SOFTWARE DOCUMENTATION NONEXISTENT. PLEASE GET IN TOUCH IF YOU NEED HELP**

Installation and Usage
----------------------

The software is **not a package** and cannot be installed by `devtools::install_github` or related. Sorry. The simplest way to get the software is to download or clone the entire repo.

The code mostly uses base R capabilities but has a few dependencies: `RColorBrewer`, `akima`, `knitr`, and `pryr`. Since it's not a package, you have to manually install these packages if you don't already have them.

The recommended way to run the program is to `source` the file `R/run.R` into your R session; `R/run.R` will source the rest. Once loaded, you can run the program by executing the statement `run()` as shown below.

``` r
## This code block assumes your working directory is the root of the distribution

source('R/run.R');
run();
```

This runs the program in a demo-like mode that quickly generates the data and produces the figures and table that appear in this README document. The default computation runs three kinds of simulations totaling 47,000 instances, generates 10 other small data files, and produces 8 figures and one table. The data generation takes about 11 seconds on my small Linux server; the figures and table take about 20 seconds, much of which is spent rendering the plots over a remote X11 connection.

The code that generates the data is in `R/dat_readme.R`; the code that creates the figures and table is in `R/doc_readme.R`.

You can run each part separately by running one or more of the statements below.

``` r
## This code block assumes your working directory is the root of the distribution
## and you've already sourced R/run.R into your session

init(doc='readme');          # you MUST specify doc to run the program in pieces
dodat();                     # generate data
dodoc();                     # generate figures and tables
dodoc(save.out=F);           # generate figures and tables without saving them
dodoc(figscreen=F);          # generate and save figures without plotting to screen. much faster!
```

The program can also generate the data and outputs for the documents associated with the project. To generate these, execute `run()` with a suitable `doc` argument as shown below.

``` r
## This code block assumes your working directory is the root of the distribution.

source('R/run.R');
run(doc='ovrfx');            # run code for blog post on significant effect size inflation
```

The `doc` arguments for each document are

<table style="width:75%;">
<colgroup>
<col width="37%" />
<col width="37%" />
</colgroup>
<thead>
<tr class="header">
<th>document</th>
<th><code>doc</code> argument</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="https://replicationnetwork.com/2019/02/16/goodman-when-youre-selecting-significant-findings-youre-selecting-inflated-estimates/">When You Select Significant Findings, You’re Selecting Inflated Estimates</a></td>
<td><code>ovrfx</code></td>
</tr>
<tr class="even">
<td><a href="https://replicationnetwork.com/2019/05/01/your-p-values-are-too-small-and-so-are-your-confidence-intervals/">Your P-Values are Too Small! And So Are Your Confidence Intervals!</a></td>
<td><code>ovrht</code></td>
</tr>
<tr class="odd">
<td><a href="https://natgoodman.github.io/misig">README</a></td>
<td><code>readme</code></td>
</tr>
<tr class="even">
<td><a href="https://natgoodman.github.io/misig/readmesupp.stable.html">Supplementary Material for README</a></td>
<td><code>readmesupp</code></td>
</tr>
<tr class="odd">
<td>supplementary material for blog posts TBD</td>
<td>TBD</td>
</tr>
</tbody>
</table>

Nomenclature
------------

<table>
<colgroup>
<col width="12%" />
<col width="12%" />
<col width="74%" />
</colgroup>
<thead>
<tr class="header">
<th>in code</th>
<th>in text</th>
<th>meaning</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>n</code></td>
<td><span class="math inline"><em>n</em></span></td>
<td>sample size</td>
</tr>
<tr class="even">
<td><code>d.pop</code></td>
<td><span class="math inline"><em>d</em><sub><em>p</em><em>o</em><em>p</em></sub></span></td>
<td>population effect size</td>
</tr>
<tr class="odd">
<td><code>d.sdz</code></td>
<td><span class="math inline"><em>d</em><sub><em>s</em><em>d</em><em>z</em></sub></span></td>
<td>standardized observed effect size, aka Cohen's d</td>
</tr>
<tr class="even">
<td><code>d.het</code></td>
<td><span class="math inline"><em>d</em><sub><em>h</em><em>e</em><em>t</em></sub></span></td>
<td>mean of heterogeneous population distribution</td>
</tr>
<tr class="odd">
<td><code>sd.het</code></td>
<td><span class="math inline"><em>s</em><em>d</em><sub><em>h</em><em>e</em><em>t</em></sub></span></td>
<td>standard deviation of heterogeneous population distribution</td>
</tr>
<tr class="even">
<td><code>d</code></td>
<td><span class="math inline"><em>d</em></span></td>
<td>variously means either <span class="math inline"><em>d</em><sub><em>p</em><em>o</em><em>p</em></sub></span>, <span class="math inline"><em>d</em><sub><em>s</em><em>d</em><em>z</em></sub></span>, or <span class="math inline"><em>d</em><sub><em>h</em><em>e</em><em>t</em></sub></span>; hopefully clear from context</td>
</tr>
<tr class="odd">
<td><code>meand.simu</code></td>
<td><span class="math inline"><em>m</em><em>e</em><em>a</em><em>n</em><em>d</em><sub><em>s</em><em>i</em><em>m</em><em>u</em></sub></span></td>
<td>mean significant observed effect size computed from simulated data</td>
</tr>
<tr class="even">
<td><code>meand.theo</code></td>
<td><span class="math inline"><em>m</em><em>e</em><em>a</em><em>n</em><em>d</em><sub><em>t</em><em>h</em><em>e</em><em>o</em></sub></span></td>
<td>mean significant observed effect size computed analytically</td>
</tr>
<tr class="odd">
<td><code>pval.simu</code></td>
<td><span class="math inline"><em>p</em><em>v</em><em>a</em><em>l</em><sub><em>s</em><em>i</em><em>m</em><em>u</em></sub></span></td>
<td>p-value computed from simulated data</td>
</tr>
<tr class="even">
<td><code>pval.theo</code></td>
<td><span class="math inline"><em>p</em><em>v</em><em>a</em><em>l</em><sub><em>t</em><em>h</em><em>e</em><em>o</em></sub></span></td>
<td>p-value computed analytically</td>
</tr>
</tbody>
</table>

<br/>

The program performs three types of simulations.

1.  *fixd*. Fixes *d*<sub>*p**o**p*</sub> to a few values of interest, sets *n* to a range of values, and simulates many studies for each *d*<sub>*p**o**p*</sub> and *n*.
2.  *rand*. Randomly selects a large number of *d*<sub>*p**o**p*</sub>s, sets *n* to a range of values, and simulates one study for each *d*<sub>*p**o**p*</sub> and *n*.
3.  *hetd*. Sets *d*<sub>*h**e**t*</sub> and *s**d*<sub>*h**e**t*</sub> to a few values of interest, and selects a large number of *d*<sub>*p**o**p*</sub>s from normal distributions with mean *d*<sub>*h**e**t*</sub> and standard deviation *s**d*<sub>*h**e**t*</sub>. Then sets *n* to a range of values, and simulates one study for each *d*<sub>*p**o**p*</sub> and *n*.

The program implements two data generation models:

1.  *fixed effect (fix)* which imagines we're repeating a study an infinite number of times, drawing a different sample each time from a population with a fixed true effect size
2.  *heterogeneous effect (het)* which assumes that each time we do the study, we’re sampling from a different population with a different true effect size

Figures and Tables
------------------

The default mode produces figures that illustrate the kinds of graphs the program can produce.

1.  *d* vs. *d* scatter plot colored by p-value; by default *d*<sub>*p**o**p*</sub> vs. *d*<sub>*s**d**z*</sub>
2.  histogram typically of *d*<sub>*s**d**z*</sub> colored by p-value
3.  probability distribution vs. *d*<sub>*s**d**z*</sub> colored by p-value; can compute probability density or cumulative probability internally, or you can pass the distribution into the function
4.  multiple line plot; my adaptation of R's `matplot`

The first figure shows a *d* vs. *d* scatter plot of *rand* simulated data for *n* = 20. The second uses the same data but zooms in to the critical region where p-values switch from nonsignificant to significant.

<img src="figure/readme/figure_001a_plotdvsd.png" width="50%" /><img src="figure/readme/figure_001b_plotdvsd_zoom.png" width="50%" />


NEWS.md
Revision history for misig repository
================
Nathan (Nat) Goodman
December 1, 2019

<!-- NEWS.md is generated from NEWS.Rmd. Please edit that file -->
Release 1.40 2019-12-01
-----------------------

Support posted version of `confi` blog post. 'Ladies And Gentlemen, I Introduce To You, “Plausibility Limits”'.

**CAVEAT:** TRN needs Word documents, but the Rmd-to-Word converter doesn't handle output scaling of figures. The workaround I devised for the `effit` post - plotting mutliple figures to the same "device" (output file) - didn't work here: the scaling messed up the in-plot text. Since the scaling didn't work, I didn't bother learning how to adjust the file layout for for 1 row, 2 columns. Instead, I used the same full-size figures as in the html and pdf versions, and manually scaled the figures in Word to 40% scale.

Major code changes

-   Added `withrows` function to iterate over rows of `cases` data frame. Replaces `withcase`

New files

-   `confi.Rmd` and related
    -   Source files for `confi` blog post.
-   `dat_confi.R`, `doc_confi.R`
    -   Data and document generation functions for `confi` blog post

Changed files

-   `R/dat.R`
    -   Do simulations in chunks. Much more efficient when `m` is big
    -   Use `withrows` instead of `withcase` to iterate over cases
-   `dat_ovrfx.R`, `doc_ovrfx.R`, `dat_ovrht.R`, `doc_ovrht.R`, `dat_readme.R`, `doc_readme.R`
    -   Minor parameter changes mainly for stylistic consistency
-   `dat_readmesupp.R`
    -   Minor parameter changes mainly for stylistic consistency
-   `doc_readmesupp.R`
    -   Use `withrows` instead of `withcase` to iterate over cases
    -   Minor parameter changes mainly for stylistic consistency
-   `R/datman.R`
    -   Use `nvq_file` to construct filenames more flexibly. Especially useful for simulation chunks
    -   Add function to concentate simulation chunks
-   `R/doc.R`
    -   Add support for multiple figures per plot file (`figdev`). Used to generare nicer Word output for TRN
    -   Continue replacing deprecated `parent` function by `param`
-   `R/init.R`
    -   Add `tmpdir`. Used mainly for simulation chunks
    -   Only create output directories needed by specific doc
    -   Remove params for obsolete `supp` documents
    -   Add ability to clean specific data types: subdirs and top level files
-   `R/plot.R`
    -   Add `legend.label` to top-level plot functions to control label for `pval_legend`. Used to display plausibility legends in `confi` blog post
    -   Add `plot_empty` function. Was used in now obsolete `ovrhtsupp`. Though not used now, seems worth keeping
    -   Add `fill_area` function to fill abitrary areas
    -   `plotpvsd`: simplify setting of line properties for `confi` blog post
    -   `plotm`: let `x` be matrix; add `linear` smoothing method
    -   `vhline`: generalize `hlab`, `vlab` controls for displaying labels
    -   `pval_legend`: add `label` param to control legend title
    -   `fill-tail`: continue removing deprecated `parent` function
-   `R/source.R`
    -   Add `source_ifexists` function to `source` optional files
    -   Only `source` doc-specific files if `param` environment setup; fixes bug in *de novo* startup
-   `R/stats.R`
    -   Change `ci.lo`, `ci.hi` variables to `lo`, `hi` for style
-   `R/util.R`
    -   Upper-case argument names in functions like `nvq` to avoid conflicts with user-supplied arguments
    -   Add `nvq_file` function to more flexibly generate filenames
    -   Add `i_pretty` function to pretty print `i` (used for simulation chunks)
    -   Add `pretty` wrapper for pretty printing functions
    -   Continue trying to create `parent` function that works :). Not sure I have it yet...
    -   Add `linear` smoothing method
    -   Add `smooth` wrapper for smoothing functions
    -   Add `span` parameter to `loessm` smoothing function
    -   Add `withrows` function to iterate over `cases` data frame; replaces `withcase`
    -   Add rounding functions
    -   Add `pick` function to pick n items approximately evenly spaced
    -   Add `tol` argument to `between` to handle imprecise decimals
    -   Add `tback` function: wrapper for `traceback` with arguments set the way I like them

Release 1.30 2019-06-12
-----------------------

First real version of README supplement

Major code changes

-   Improve code for figure titles: generic `figtitle` function to generate figure titles instead of doc-specific functions; auto-scaling instead of setting `cex.main`
-   Start cleanup of figure/table numbering scheme

New files

-   `READMEsupp.Rmd`
    -   Source file for README supplement
-   `R/dat_readmesupp.R`, `R/doc_readmesupp.R`

R/plot.R
#################################################################################
##
## Author:  Nat Goodman
## Created: 19-01-09
##          uses code from repwr/R/plot.R created 18-05-03
##
## Copyright (C) 2019 Nat Goodman.
## 
## Plotting code for effit documents
##
## This software is open source, distributed under the MIT License. See LICENSE
## file at https://github.com/natgoodman/NewPro/FDR/LICENSE 
##
#################################################################################

## ---- Plot Functions ----
## plot d.y vs d.x, typically d.pop vs d.sdz, colored by pval
## sim is data frame of simulation results
## params for generating pval colors
##   n is sample size - default from sim
##   sd.het is sd for noncentral d2ht distribution - default from sim if exists there
##   distribution is d2t or d2ht - default 'd2t' unless sd.het is set
## title is title
## cex.title is what R calls cex.main. if 'auto' or NULL, use auto-scaling
## legend tells whether to draw pval legend
##   legend.xscale, legend.yscale are fractions of plat ares
##   legend.cex is cex for legend text
## x, y are columns of sim containing x, y values; also used for axis labels
## vline,hline are vectors of x or y positions for extra vertical or horizontal lines
## vhlty, vhcol, vhlwd are lty, col, lwd for these extra lines
## vlab, hlab contol writing vline, hline values along axes
## vhdigits is number of digits for these values
#
plotdvsd=
  function(sim,
           n=unique(sim$n),sd.het=if(exists('sd.het',sim))unique(sim$sd.het) else NULL,
           distribution=cq(d2t,d2ht),
           title='',cex.title='auto',x='d.sdz',y='d.pop',
           col=NULL,
           legend=TRUE,legend.xscale=1/8,legend.yscale=1/3,legend.x0=NULL,
           legend.label='p-value',legend.cex=0.75,
           vline=NULL,hline=NULL,vhlty='dashed',vhcol='grey50',
           vhlwd=1,vlab=TRUE,hlab=TRUE,vhdigits=2,
           xlab=switch(x,
                       d.sdz="observed effect size (Cohen's d)",
                       d.pop="true effect size",
                       NULL),
           ylab=switch(y,
                       d.sdz="observed effect size (Cohen's d)",
                       d.pop="true effect size",
                       NULL),
           ...) {
    if (nrow(sim)==0) {
      warning('sim is empty. nothing to plot');
      return();
    }
    if (is.null(col)) {
      if (missing(distribution)) distribution=if(is.null(sd.het)) 'd2t' else 'd2ht'
      else distribution=match.arg(distribution);
      if (length(n)==1&length(sd.het)<=1)
        col=d2col(n=n,sd.het=sd.het,distribution=distribution,d=sim$d.sdz)
      else {
        if (length(n)!=1) warning('plotdvsd needs unique n to convert d to color');
        if (length(sd.het)>1) warning('plotdvsd needs unique sd.het to convert d to color');
        col='grey';
      }}
    if (is.null(cex.title)|cex.title=='auto') cex.title=cex_title(title);
    plot(sim[,x],sim[,y],col=col,main=title,cex.main=cex.title,xlab=xlab,ylab=ylab,pch=19,
         cex=0.5,...);
    grid();
    ## plot extra lines & values if desired. nop if vline, hline NULL
    vhline(vline=vline,hline=hline,vlab=vlab,hlab=hlab,vhdigits=vhdigits,
           lty=vhlty,col=vhcol,lwd=vhlwd);
    ## plot legend if desired
    if (legend)
      pval_legend(x.scale=legend.xscale,y.scale=legend.yscale,x0=legend.x0,label=legend.label,
                  cex=legend.cex);
  }
## plot histogram of (typically) d.sdz colored by pval
## sim is data frame of simulation results
## params for generating pval colors
##   n is sample size - default from sim
##   sd.het is sd for noncentral d2ht distribution - default from sim if exists there
##   distribution is d2t or d2ht - default 'd2t' unless sd.het is set
## title is title
## legend tells whether to draw pval legend
##   legend.x0 is harcoded x-position of legend - use with care!
##   legend.xscale, legend.yscale are fractions of plat areas
##   legend.cex is cex for legend text
## x is column of sim containing values for histogram; also used for axis labels
## breaks, freq are same-named parameters for hist
## vline,hline are vectors of x or y positions for extra vertical or horizontal lines
## vhlty, vhcol, vhlwd are lty, col, lwd for these extra lines
## vlab, hlab contol writing vline, hline values along axes
## vhdigits is number of digits for these values
#
plothist=
  function(sim,
           n=unique(sim$n),sd.het=if(exists('sd.het',sim))unique(sim$sd.het) else NULL,
           distribution=cq(d2t,d2ht),
           title='',cex.title='auto',x='d.sdz',breaks=50,freq=FALSE,add=FALSE,
           col=NULL,border='black',
           legend=TRUE,legend.xscale=1/8,legend.yscale=1/3,legend.x0=NULL,
           legend.label='p-value',legend.cex=0.75,
           vline=NULL,hline=NULL,vhlty='dashed',vhcol='grey50',
           vhlwd=1,vlab=TRUE,hlab=TRUE,vhdigits=2,
           xlab="observed effect size (Cohen's d)",ylab="density",
           ...) {
   if (nrow(sim)==0) {
      warning('sim is empty. nothing to plot');
      return();
   }
   hist.obj=hist(sim[,x],breaks=breaks,plot=FALSE);
   if (is.null(col)) {
     if (missing(distribution)) distribution=if(is.null(sd.het)) 'd2t' else 'd2ht'
     else distribution=match.arg(distribution);
     if (length(n)==1&length(sd.het)<=1)
       col=d2col(n=n,sd.het=sd.het,distribution=distribution,d=hist.obj$mids)
     else {
       if (length(n)!=1) warning('plothist needs unique n to convert d to color');
       if (length(sd.het)>1) warning('plothist needs unique sd.het to convert d to color');
       col='grey';
     }}
   if (is.null(cex.title)|cex.title=='auto') cex.title=cex_title(title);
    plot(hist.obj,col=col,border=border,freq=freq,main=title,cex.main=cex.title,
         xlab=xlab,ylab=ylab,add=add,...);
    if (!add) {
      box(); grid();
    } 
    ## plot extra lines & values if desired. nop if vline, hline NULL
   vhline(vline=vline,hline=hline,vlab=vlab,hlab=hlab,vhdigits=vhdigits,
          lty=vhlty,col=vhcol,lwd=vhlwd);
    ## plot legend if desired
   if (legend)
     pval_legend(x.scale=legend.xscale,y.scale=legend.yscale,x0=legend.x0,label=legend.label, cex=legend.cex);
  }
## plot probability distributions vs. d colored by pval
## n is sample size (for converting d to t or pval)
## d, col, lwd - usually missing - can be used when defaults not desired
## lwd.sig, lwd.nonsig - lwd for sig vs. non sig
##   if lwd set, use for both 
## dlim is range of d.sdz
##   dinc is step-size for extending d across range
## d0 is center for noncentral d2t distribution
## d.het is center for noncentral d2ht distribution
## sd.het is sd for noncentral d2ht distribution
## distribution is d2t or d2ht
## y is probability density or cumulative prob. can be values or keyword
## fill.tail tells whether to fill the distribution tail (density only)
##   boolean or one or more of 'upper','lower','both'. TRUE means c('upper','lower')
## add tells whether to add new plot to existing one
## xunit tells what to write along x-axis: d, t, or both
## title is title
## legend tells whether to draw pval legend
##   legend.xscale, legend.yscale are fractions of plat areas
##   legend.cex is cex for legend text
## vline,hline are vectors of x or y positions for extra vertical or horizontal lines
## vhlty, vhcol, vhlwd are lty, col, lwd for these extra lines
## vlab, hlab contol writing vline, hline values along axes
## vhdigits is number of digits for these values
plotpvsd=
  function(n,d0=NULL,d.het=NULL,sd.het=NULL,distribution=cq(d2t,d2ht),y=cq(density,cumulative),
           d,col,
           lty.sig='solid',lty.nonsig=lty.sig,
           lwd=NULL,
           lwd.sig=if(!is.null(lwd)) lwd else 4,lwd.nonsig=if(!is.null(lwd)) lwd else lwd.sig/2,
           dlim=c(-2,2),dinc=.005,
           add=FALSE,fill.tail=FALSE,
           title='',cex.title='auto',d.crit=d_crit(n),
           legend=!add,legend.xscale=1/8,legend.yscale=1/3,legend.x0=NULL,
           legend.label='p-value',legend.cex=0.75,
           vline=NULL,hline=NULL,vhlty='dashed',vhcol='grey50',
           vhlwd=1,vlab=TRUE,hlab=TRUE,vhdigits=2,
           xlab="observed effect size (Cohen's d)",ylab="probability",
           ...) {
    if (missing(d)) d=seq(min(dlim),max(dlim),by=dinc);
    if (missing(distribution)) {
      distribution=if(is.null(d.het)&is.null(sd.het)) 'd2t' else 'd2ht';
    }
    else distribution=match.arg(distribution);
    if (mode(y)=='character') {
      y=match.arg(y);
      if (y!='density') fill.tail=FALSE;
      if (missing(ylab)) ylab=if(y=='density') 'probability density' else 'cumulative probability';
      if (distribution=='d2t') {
        y=if(y=='density') d_d2t(n,d,d0) else p_d2t(n,d,d0);
      }
      else {
        y=if(y=='density') d_d2ht(n,d.het=d.het,sd.het=sd.het,d=d)
          else p_d2ht(n,d.het=d.het,sd.het=sd.het,d=d);
      }}
    if (is.null(cex.title)|cex.title=='auto') cex.title=cex_title(title);
    if (distribution=='d2t') pval=d2pval(n,d) else pval=d2htpval(n,sd.het,d);
    if (missing(col)||is.null(col)) col=pval2col(pval);
    lty=ifelse(pval<=0.05,lty.sig,lty.nonsig);
    lwd=ifelse(pval<=0.05,lwd.sig,lwd.nonsig);
    if (is.null(cex.title)|cex.title=='auto') cex.title=cex_title(title);
    l=length(d);
    if (!add) {
      plot(d,y,type='n',xlab=xlab,ylab=ylab,main=title,cex.main=cex.title,...);
      grid();
    }
    if (l==1) points(d,y,col=col,pch=19)
    else {
      x0=d[-l]; y0=y[-l]; x1=d[-1]; y1=y[-1];
      segments(x0,y0,x1,y1,col=col,lty=lty,lwd=lwd)
    }
    ## grid();
    ## plot extra lines & values if desired. nop if vline, hline NULL
    vhline(vline=vline,hline=hline,vlab=vlab,hlab=hlab,vhdigits=vhdigits,
           lty=vhlty,col=vhcol,lwd=vhlwd);
    ## fill tail if desired. already made sure we're doing density
    if (is.logical(fill.tail)&fill.tail) fill.tail=cq(upper,lower);
    if (!is.logical(fill.tail)) {
      if ('both' %in% fill.tail) fill.tail=cq(upper,lower);
      sapply(fill.tail,function(tail) fill_tail(tail,n,d,d0,d.crit));
    }
    ## plot legend if desired
    if (legend)
      pval_legend(x.scale=legend.xscale,y.scale=legend.yscale,x0=legend.x0,label=legend.label,
                  cex=legend.cex);
  }
## plot multiple lines - my adaptation of matplot - adapted from repwr/plotratm
## x is vector or matrix of x values
## y is vector or matrix of y values
##   like matplot, each line is column of x or y
##   unlike matplot, code currently assumes at most one of x,y is matrix - for smoothing to work
## col, lty, lwd are the usual line properties
## title, cex.title are title and cex for title
## xaxt, yaxt control labels on axes - NOT YET IMPLEMENTED
##   's' or NULL means let R do it
##    else list of axis params, eg, at, labels
##           title='',cex.title='auto',
## vline,hline are vectors of x or y positions for extra vertical or horizontal lines
## vhlty, vhcol, vhlwd are lty, col, lwd for these extra lines
## vlab, hlab contol writing vline, hline values along axes
## vhdigits is number of digits for these values
## smooth whether to smooth data to make plot prettier
##   aspline, spline, loess, none, TRUE, FALSE. default is aspline.
##   TRUE means aspline. FALSE means none
##   spar is for spline
## smooth.xy tells which axis is domain of smoothing
##   default: 'x' if both x and y are vector-like, else whichever is vector-like
## legend tells whether and where to draw legend
##   TRUE or word like 'right' means draw legend, FALSE or NULL means no legend
## legend.title is legend title
## legend.args are further legend params
plotm=
  function(x,y,col='black',lty='solid',lwd=1,title='',cex.title='auto',
           xaxt='s',yaxt='s',
           vline=NULL,hline=NULL,vhlty='dashed',vhcol='grey50',
           vhlwd=1,vlab=TRUE,hlab=TRUE,vhdigits=2,
           smooth=c(cq(aspline,spline,loess,linear,none),TRUE,FALSE),smooth.xy=cq(x,y),
           spar=NULL,span=0.75,
           legend=if(is.vector(y)) FALSE else 'right',legend.title=NULL,
           legend.labels=if(is.vector(y)) NULL else colnames(y),
           legend.args=list(where=NULL,x=NULL,y=NULL,cex=0.8,
                            title=legend.title,labels=legend.labels,col=col,lty=lty,lwd=lwd),
           ...) {
    if (is.null(x)) stop("Nothing to plot: 'x' is NULL");
    if (is.null(y)) stop("Nothing to plot: 'y' is NULL");
    if (is.vector(x)) x=as.matrix(x)
    else if (length(dim(x))!=2) stop("'x' must be vector or 2-dimensional matrix-like object");
    if (is.vector(y)) y=as.matrix(y)
    else if (length(dim(y))!=2) stop("'y' must be vector or 2-dimensional matrix-like object");
    if (nrow(x)!=nrow(y)) stop("'x' and 'y' have different number of rows");
    if (is.null(cex.title)|cex.title=='auto') cex.title=cex_title(title);
    smooth=if(is.logical(smooth)&&!smooth) 'none' else smooth;
    if (smooth!='none') {
      ## at most one of x,y can be matrix-like for smoothing to work
      if (dim(x)[2]>1&dim(y)[2]>1)
        stop("At most one of 'x' or 'y' can have multiple columns for smoothing to work");
      if (missing(smooth.xy)) smooth.xy=if(dim(x)[2]==1) 'x' else 'y';
      ## smooth
      if (smooth.xy=='x') {
        x.smooth=seq(min(x),max(x),len=100);
        y=smooth(x,y,xout=x.smooth,method=smooth,spar=spar,span=span);
        x=x.smooth;
      } else {
        y.smooth=seq(min(y),max(y),len=100);
        x=smooth(x=y,y=x,xout=y.smooth,method=smooth,spar=spar,span=span);
        y=y.smooth;
      }
    }
    matplot(x,y,type='l',main=title,cex.main=cex.title,col=col,lty=lty,lwd=lwd,
            xaxt=xaxt,yaxt=yaxt,...);
    grid();
    ## plot extra lines & values if desired. nop if vline, hline NULL
    vhline(vline=vline,hline=hline,vlab=vlab,hlab=hlab,vhdigits=vhdigits,
           lty=vhlty,col=vhcol,lwd=vhlwd);
    ## draw legend if desired
    if (is.null(legend)) legend=FALSE
    else if (!is.logical(legend)) {
      legend.args$where=legend;
      legend=TRUE;
    }
    if (legend) do.call(plotm_legend,legend.args);
  }
## empty plot -kust title & axes
plotempty=
  function(title='',cex.title='auto',xlab='x',ylab='y',xlim=c(0,1),ylim=c(0,1),
           xaxp=c(xlim,1),yaxp=c(ylim,1),...) {
    if (is.null(cex.title)|cex.title=='auto') cex.title=cex_title(title);
    plot(x=NULL,y=NULL,type='n',main=title,cex.main=cex.title,xlab=xlab,ylab=ylab,
         xlim=xlim,ylim=ylim,xaxp=xaxp,yaxp=yaxp,...);
    xylim=par('usr');                   # limits of disply region
    xmid=mean(xylim[1:2]);
    ymid=mean(xylim[3:4]);
    text(xmid,ymid,'PLOT DELIBERATELY LEFT BLANK',adj=c(0.5,0.5));
    invisible();
}

## helper functions to plot horizontal and vertical line segments
vhline=function(vline=NULL,hline=NULL,vlab=TRUE,hlab=TRUE,vhdigits=2,col=NA,...) {
  xylim=par('usr');
  vline=vline[which(between(vline,xylim[1],xylim[2]))];
  hline=hline[which(between(hline,xylim[3],xylim[4]))];
  abline(v=vline,h=hline,col=col,...);
  ## write vhline values along axes
  vline=vline[vlab];
  if (length(vline)>0)
    mtext(round(vline,vhdigits),side=1,at=vline,col=col,line=0.25,cex=0.75);
  hline=hline[hlab];
  if (length(hline)>0)
    mtext(round(hline,vhdigits),side=2,at=hline,col=col,line=0.25,cex=0.75);
}
hline=
  function(y,x0=0,x,col='black',lty='solid',lwd=1,cex=0.75,text=NULL,
           label=list(text=text,side=2,at=y,col=col,line=0.25,cex=cex,las=1)) {
    segments(x0=x0,x1=x,y0=y,y1=y,col=col,lty=lty,lwd=lwd);
    if (!is.null(text)) do.call(mtext,label);
  }
vline=
  function(x,y0=0,y,col='black',lty='solid',lwd=1,cex=0.75,text=NULL,
           label=list(text=text,side=1,at=x,col=col,line=0.25,cex=cex,las=1)) {
    segments(x0=x,x1=x,y0=y0,y1=y,col=col,lty=lty,lwd=lwd);
    if (!is.null(text)) do.call(mtext,label);
  }

## draw pval legend. works for big picture figure and probability plots
pval_legend=function(x.scale,y.scale,x0=NULL,cex,label='p-value') {
  param(brk.pval,col.pval,steps.pvcol,sig.level);
  ## plt=par('usr');                       # plot region in user coordinates
  ## names(plt)=cq(left,right,bottom,top);
  xtkl=par('xaxp');                    # x tick locations
  ytkl=par('yaxp');                    # y tick locations
  names(xtkl)=names(ytkl)=cq(lo,hi,num);
  if (is.null(x0)) x0=xtkl['lo'];
  width=x.scale*(xtkl['hi']-x0);
  x1=x0+width;
  y1=ytkl['hi'];
  height=y.scale*(y1-ytkl['lo']);
  y0=y1-height;
  ## image sometimes leaves blank space when y0 is between tick marks. roundoff problem, I think
  ## works better to stretch y0 to next lower tick
  tkl=seq(ytkl['lo'],y1,len=ytkl['num']+1);
  y0=tkl[findInterval(y0,tkl)];
  height=y1-y0;                         # adjust height for new y0
  x=c(x0,x1);
  y=seq(y0,y1,length.out=2*steps.pvcol+1)[1:(2*steps.pvcol)]
  z=t(as.matrix(rev(head(brk.pval,-1))));
  image(x,y,z,add=TRUE,breaks=brk.pval,col=col.pval);
  ## add legend text
  x1=x1+strwidth(' ',cex=cex);
  text(x1,y0,0,adj=c(0,0),cex=cex)
  text(x1,y0+height/2,sig.level,adj=c(0,0.5),cex=cex)
  text(x1,y0+height,1,adj=c(0,1),cex=cex)
  y1=y1+strheight('p-value',cex=cex);
  text(x0+width/2,y1,label,adj=c(0.5,0.5),cex=cex);
}
## draw plotm legend. adapted from repwr/mess_legend
plotm_legend=
  function(where=NULL,x=NULL,y=NULL,cex=0.8,bty='n',
           title=NULL,title.col='black',
           col='black',lty='solid',lwd=1,labels=NULL,...) {
    if (is.null(labels)) return();      # nothing to draw
    if (is.null(x)) x=where;
    legend(x,y,bty=bty,legend=labels,cex=cex,col=col,lwd=lwd,lty=lty,
          title=title,title.col=title.col,...);
  }

## fill tail of probability density
## adapted from https://stackoverflow.com/questions/45527077. Thx!
fill_tail=function(tail=cq(upper,lower),n,d,d0,d.crit) {
  tail=match.arg(tail);
  x=if(tail=='upper') x=d[d>d.crit] else x=d[d<(-d.crit)]
  y=d_d2t(n,x,d0);
  ## toss in extra x, y values to close polygon
  x=c(min(x),x,max(x));
  y=c(0,y,0);
  ## looks nicer if y stops just short of curve
  y=sapply(y,function(y) max(0,y-1e-3));
  ## do it!
  polygon(x=x,y=y,col='grey',border=NA);
}
## TODO: y.gap doesn't really do what I want...
fill_area=function(n,d0,d,col='grey',y.gap=1e-3) {
  y=d_d2t(n,d,d0);
  ## toss in extra x, y values to close polygon
  x=c(min(d),d,max(d));
  y=c(0,y,0);
  ## looks nicer if y stops just short of curve
  y=sapply(y,function(y) max(0,y-y.gap));
  ## do it!
  polygon(x=x,y=y,col=col,border=NA);
}

pval2col=function(pval) {
  param(col.pval,brk.pval,min.pvcol);
  col.pval[findInterval(-log10(clamp_pval(pval,min.pvcol)),brk.pval,all.inside=TRUE)];
}
d2col=function(n,sd.het,distribution,d) {
  pval=if(distribution=='d2t') d2pval(n,d) else d2htpval(n,sd.het,d);
  pval2col(pval);
}
clamp_pval=function(pval,min.pvcol) sapply(pval,function(pval) max(min(pval,1),min.pvcol));

## auto-scale title
cex_title=function(title) {
  xyplt=par('plt');                     # dimensions of plot region
  xplt=xyplt[2]-xyplt[1];               # width of plot region
  min(1,xplt/strwidth(title,units='fig'));
}
